### PR TITLE
prefetch required stdin arguments in CheckArguments

### PR DESCRIPTION
**Pull request by Stebalien**
Also:

* Make it easier to actually use ParseBodyArguments. Now it only returns an error when something actually goes wrong.
* Add a function for validating a command tree (to be run from a test case).

### CODE REVIEW

**Comment by Stebalien** [comment]
(tested with go-ipfs and it fixes the issue in question)

File: command.go [hunk 1]
```diff
@@ -162,55 +164,120 @@ func (c *Command) GetOptions(path []string) (map[string]cmdkit.Option, error) {
...
+		return nil
+	}
 
-	// count required argument definitions
-	numRequired := 0
-	for _, argDef := range c.Arguments {
-		if argDef.Required {
-			numRequired++
+	lastArg := c.Arguments[len(c.Arguments)-1]
+	if req.bodyArgs == nil &&
```
Comment (whyrusleeping) [inline command.go]: maybe add a "TODO: avoid calling this function twice"

Comment (Stebalien) [reply command.go]: Filed an issue.

File: command.go [hunk 2]
```diff
@@ -162,55 +164,121 @@ func (c *Command) GetOptions(path []string) (map[string]cmdkit.Option, error) {
...
+		lastArg.SupportsStdin &&
+		lastArg.Type == cmdkit.ArgString &&
+		req.Files != nil {
+
+		fi, err := req.Files.NextFile()
+		switch err {
+		case io.EOF:
+		case nil:
+			req.bodyArgs = bufio.NewScanner(fi)
+			// Can't pass files and stdin arguments.
```
Comment (Kubuxu) [inline command.go]: This mean that `ipfs add ./file -` is not possible?

File: command.go [hunk 3]
```diff
@@ -162,55 +164,121 @@ func (c *Command) GetOptions(path []string) (map[string]cmdkit.Option, error) {
...
+	if len(c.Arguments) == 0 {
+		return nil
+	}
 
-	// count required argument definitions
-	numRequired := 0
-	for _, argDef := range c.Arguments {
-		if argDef.Required {
-			numRequired++
+	lastArg := c.Arguments[len(c.Arguments)-1]
```
Comment (Kubuxu) [inline command.go]: Won't this panic in case of command with no arguments?

Comment (Kubuxu) [reply command.go]: Ahh, line above.

File: command.go (hunk 2 above)
Comment (Kubuxu) [reply command.go]: Ok, so this is only ArgString type.

Comment (Stebalien) [reply command.go]: Sorry, stdin *string* arguments.

File: command.go [hunk 4]
```diff
@@ -162,55 +164,121 @@ func (c *Command) GetOptions(path []string) (map[string]cmdkit.Option, error) {
...
+		lastArg.Type == cmdkit.ArgString &&
+		req.Files != nil {
+
+		fi, err := req.Files.NextFile()
+		switch err {
+		case io.EOF:
+		case nil:
+			req.bodyArgs = bufio.NewScanner(fi)
+			// Can't pass files and stdin arguments.
+			req.Files = nil
```
Comment (whyrusleeping) [inline command.go]: I don't think we need to set this to nil here.

It should be fine that we have called NextFile (popping that off the queue)

**Pull request merged**